// @gamex/cc-ecs/assets/base/BaseEntity.ts
import { Node } from 'cc';
import { EcsObject, IComponent, IEntity, IFlag, ITypeofComponent, IWorld } from '../core/EcsObject';
import { UuidMaker } from '../core/EcsUtils';

/**uuid生成器 */
export const entityUuidMaker = new UuidMaker();

/**实体基类 */
export abstract class BaseEntity extends EcsObject implements IEntity {
    public static isEntity(param: unknown): param is IEntity {
        if (!param) return false;
        return param instanceof BaseEntity;
    }

    /**内部函数 */
    public static updateUUID(entity: IEntity) {
        (entity as BaseEntity)._uuid = entityUuidMaker.create();
    }
    /**内部函数 */
    public static UpdateMatch(entity: IEntity) {
        (entity as BaseEntity).innerUpdateMatch(entity);
    }
    protected abstract innerUpdateMatch(entity: IEntity): void;

    static initialize(entity: IEntity, world: IWorld, node: Node | null = null) {
        return (entity as BaseEntity).innerInit(world, node);
    }
    protected abstract innerInit(ecs: IWorld, node?: Node | null): void;

    static addComponent(entity: IEntity, com: IComponent) {
        return (entity as BaseEntity).innerAddComponent(com);
    }
    protected abstract innerAddComponent(com: IComponent): void;

    static removeComponent(entity: IEntity, com: IComponent) {
        return (entity as BaseEntity).innerRemoveComponent(com);
    }
    protected abstract innerRemoveComponent(com: IComponent): void;

    /**唯一ID */
    private _uuid = entityUuidMaker.create();
    public get uuid() {
        return this._uuid;
    }

    /**允许回收复用(默认false) */
    static allowRecycling = false;
    /**允许回收复用(默认false) */
    public get allowRecycling() {
        return (this.constructor as typeof BaseEntity).allowRecycling;
    }

    abstract node: Node;
    abstract isValid: boolean;
    abstract enabled: boolean;
    abstract destroy(): boolean;
    abstract checkFlagAll(flag: IFlag): boolean;
    abstract checkFlagAny(flag: IFlag): boolean;
    abstract checkFlagOnly(flag: IFlag): boolean;
    abstract add<T extends ITypeofComponent>(Com: T | IComponent): InstanceType<T> | null;
    abstract has(Com: ITypeofComponent): boolean;
    abstract get<T extends ITypeofComponent>(Com: T): InstanceType<T> | null;
    abstract getAll<T extends ITypeofComponent>(Com: T): InstanceType<T>[];
    abstract remove(com: IComponent): boolean;
    abstract removeAll(Com: ITypeofComponent): boolean;
}
