// @gamex/cc-ecs/assets/base/BaseComponent.ts
import { EcsObject, IComponent, IEntity } from '../core/EcsObject';
import { UuidMaker } from '../core/EcsUtils';

export const componentUuidMaker = new UuidMaker();

/**组件基类 */
export abstract class BaseComponent extends EcsObject implements IComponent {
    public static isComponent(param: unknown): param is IComponent {
        if (!param) return false;
        return param instanceof BaseComponent;
    }

    /**内部函数 */
    public static updateUUID(comp: IComponent) {
        (comp as BaseComponent)._uuid = componentUuidMaker.create();
    }
    /**内部函数 */
    public static updateEnable(comp: IComponent) {
        (comp as BaseComponent).innerUpdateEnable();
    }
    protected abstract innerUpdateEnable(): any;
    /**内部函数 */
    static onAdd(comp: IComponent, entity: IEntity) {
        return (comp as BaseComponent).innerAdd(entity);
    }
    protected abstract innerAdd(entity: IEntity): any;
    /**内部函数 */
    static onRemove(comp: IComponent) {
        return (comp as BaseComponent).innerRemove();
    }
    protected abstract innerRemove(): any;

    /**唯一ID */
    private _uuid = componentUuidMaker.create();
    public get uuid() {
        return this._uuid;
    }

    /**允许回收复用(默认false) */
    static allowRecycling = false;
    /**允许回收复用(默认false) */
    public get allowRecycling() {
        return (this.constructor as typeof BaseComponent).allowRecycling;
    }

    /**允许在同一个实体中重复添加(默认false)  */
    static allowMultiple = false;
    /**允许在同一个实体中重复添加(默认false) */
    public get allowMultiple() {
        return (this.constructor as typeof BaseComponent).allowMultiple;
    }

    abstract entity: IEntity;
    abstract isValid: boolean;
    abstract enabled: boolean;
    abstract get enabledInHierarchy(): boolean;
    abstract destroy(): boolean;
}
