// @gamex/cc-ecs/assets/base/BaseSystem.ts
import { Filter } from '../core/EcsFilter';
import { Matcher } from '../core/EcsMatcher';
import { EcsObject, IEntity, ISystem, IWorld } from '../core/EcsObject';
import { UuidMaker } from '../core/EcsUtils';

export const systemUuidMaker = new UuidMaker();

/**系统基类 */
export abstract class BaseSystem extends EcsObject implements ISystem {
    public static isSystem(param: unknown): param is ISystem {
        if (!param) return false;
        return param instanceof BaseSystem;
    }
    /**内部函数 */
    public static initialize(system: ISystem, world: IWorld) {
        return (system as BaseSystem).initialize(world);
    }
    /**初始化 */
    protected abstract initialize(world: IWorld): any;
    /**内部函数 */
    public static onAdd(system: ISystem, matcher: Matcher) {
        (system as BaseSystem).onAdd();

        let filter = (system as BaseSystem).matcher();
        if (filter == null) return;

        let group = matcher.register(filter);
        if (group == null) return;

        group.onEntityEnter((system as BaseSystem).onEntityEnter.bind(system));
        group.onEntityLeave((system as BaseSystem).onEntityLeave.bind(system));
        group.getEntities().forEach(entity => {
            (system as BaseSystem).onEntityEnter(entity);
        });
    }
    /**添加*/
    protected abstract onAdd(): any;
    /**内部函数 */
    public static onRemove(system: ISystem) {
        (system as BaseSystem).onRemove();
    }
    /**移除*/
    protected abstract onRemove(): any;
    /**内部函数 */
    public static execute(system: ISystem, dt: number, args: any[]) {
        (system as BaseSystem).execute(dt, ...args);
    }
    /**内部函数 */
    public static beforeExecute(system: ISystem, dt: number, args: any[]) {
        (system as BaseSystem).beforeExecute(dt, ...args);
    }
    /**内部函数 */
    public static afterExecute(system: ISystem, dt: number, args: any[]) {
        (system as BaseSystem).afterExecute(dt, ...args);
    }
    /**内部函数 */
    public static update(system: ISystem, dt: number, args: any[]) {
        (system as BaseSystem).update(dt, ...args);
    }
    /**内部函数 */
    public static beforeUpdate(system: ISystem, dt: number, args: any[]) {
        (system as BaseSystem).beforeUpdate(dt, ...args);
    }
    /**内部函数 */
    public static afterUpdate(system: ISystem, dt: number, args: any[]) {
        (system as BaseSystem).afterUpdate(dt, ...args);
    }

    /**唯一ID */
    private _uuid = systemUuidMaker.create();
    public get uuid() {
        return this._uuid;
    }

    /**
     * 过滤实体
     * - 配置后才能响应onEntityEnter、onEntityLeave
     */
    protected abstract matcher(): Filter;

    /**实体进入系统，会在execute流程最开始调用*/
    protected abstract onEntityEnter(entity: IEntity): any;

    /**实体离开系统，会在execute流程最开始调用*/
    protected abstract onEntityLeave(entity: IEntity): any;

    protected abstract execute(dt?: number, ...args: any[]): any;
    protected abstract beforeExecute(dt?: number, ...args: any[]): any;
    protected abstract afterExecute(dt?: number, ...args: any[]): any;

    // 由ecs.update驱动
    protected abstract update(dt?: number, ...args: any[]): any;
    protected abstract beforeUpdate(dt?: number, ...args: any[]): any;
    protected abstract afterUpdate(dt?: number, ...args: any[]): any;
}
