// @gamex/cc-ecs/assets/ecs/EcsEntity.ts
import { error, Node } from 'cc';
import { DEV } from 'cc/env';
import { BaseComponent } from '../base/BaseComponent';
import { BaseEntity } from '../base/BaseEntity';
import { componentCache } from '../core/EcsCache';
import { classManager, ecsclass, flagManager } from '../core/EcsManager';
import { IComponent, IComponentName, IEntity, ITypeofComponent, IWorld } from '../core/EcsObject';
import { ArrayMap } from '../core/EcsUtils';
import { EcsWorld } from './EcsWorld';

@ecsclass('EcsEntity')
export class EcsEntity extends BaseEntity implements IEntity {
    /**位运算 */
    private flag: number[] = new Array(flagManager.bits).fill(0);

    private instNames: string[] = [];

    /**组件的名字 => 组件数组 */
    private instNameToComps: ArrayMap<IComponentName, IComponent> = new ArrayMap();

    /**组件父类的名字 => 组件数组 */
    private superNameToComps: ArrayMap<IComponentName, IComponent> = new ArrayMap();

    private _isValid = false;
    /**是否有效 */
    public get isValid() {
        return this._isValid;
    }
    private set isValid(value) {
        this._isValid = value;
    }

    private _enabled = true;
    /**标记是否生效 */
    public get enabled() {
        return this._enabled;
    }
    public set enabled(value) {
        if (this._enabled === value) return;
        this._enabled = value;
        if (value) {
            this.onEnable();
        } else {
            this.onDisable();
        }
        // 更新组件状态
        this.instNameToComps.forEach((comps) => {
            for (let index = 0; index < comps.length; index++) {
                BaseComponent.updateEnable(comps[index] as BaseComponent);
            }
        });
        this._world?.UpdateEntityMatch(this);
    }

    /**指定ecs */
    private _world: EcsWorld | null = null;
    /**获取对应的ecs */
    protected get world() {
        return this._world as IWorld;
    }

    /**节点 */
    private _node: Node | null = null;
    public get node() {
        return this._node;
    }
    public set node(value: Node | null) {
        this._node = value;
    }

    /**标记是否已经销毁 */
    private _isDestroyed = true;

    /**
     * 内部初始化函数
     */
    protected innerInit(ecs: IWorld, node: Node | null = null) {
        this._world = ecs as EcsWorld;
        this._node = node;
        this._isValid = true;
        this._enabled = true;
        this._isDestroyed = false;
        this.onAdd();
        this._enabled && this.onEnable();
    }

    protected innerUpdateMatch(entity: IEntity): void {
        (entity as EcsEntity)._world?.UpdateEntityMatch(this);
    }

    /**
     * 添加组件(内部调用)
     */
    protected innerAddComponent(com: BaseComponent) {
        // 组件类名字
        const componentName = com.ecsName;

        // 添加父类名字到uuid的转换
        classManager.eachSuperName(componentName, (superName) => {
            this.superNameToComps.add(superName, com);
        });

        // 添加名字到uuid的转换
        const len = this.instNameToComps.add(componentName, com);

        if (len === 1) {
            // 记录组件实例的名字
            this.instNames.push(componentName);
            // 更新标记位
            flagManager.addFlag(this.flag, flagManager.getAllByName(componentName));
        }

        BaseComponent.onAdd(com, this);
    }

    /**
     * 移除组件(内部调用)
     */
    protected innerRemoveComponent(com: BaseComponent) {
        // 组件类名字
        const componentName = com.ecsName;

        // 删除父类名字到uuid的转换
        classManager.eachSuperName(componentName, (superName) => {
            this.superNameToComps.sub(superName, com);
        });

        // 删除名字到uuid的转换
        const del = this.instNameToComps.subDel(componentName, com);

        if (del) {
            const index = this.instNames.indexOf(componentName);
            this.instNames.splice(index, 1);
            // 更新标记位
            flagManager.getAllByNames(this.instNames, this.flag);
        }

        BaseComponent.onRemove(com);
    }

    /**
     * 根据组件名字获取对应的组件数组(先查inst再查询super)
     */
    private getCompsByName(componentName: string): readonly IComponent<IEntity>[] | null {
        const instComps = this.instNameToComps.get(componentName);
        if (instComps && instComps.length) {
            return instComps;
        }

        const superComps = this.superNameToComps.get(componentName);
        if (superComps && superComps.length) {
            return superComps;
        }

        return null;
    }

    /**
     * 根据组件名字获取对应的组件数组(先查inst再查询super)
     */
    private getCompsByName2(componentName: string) {
        const out: IComponent[] = [];
        const instComps = this.instNameToComps.get(componentName);
        if (instComps && instComps.length) {
            Array.prototype.push.apply(out, instComps);
        }

        const superComps = this.superNameToComps.get(componentName);
        if (superComps && superComps.length) {
            Array.prototype.push.apply(out, superComps);
        }

        return out;
    }

    /**
     * 移除当前所有组件
     */
    private clear() {
        const keys = this.instNames;

        for (let index = keys.length - 1; index >= 0; index--) {
            const comps = this.instNameToComps.get(keys[index]);
            if (!comps) continue;// 如果在某个组件的onRemove中主动移除其它组件，这个会是null
            for (let index = comps.length - 1; index >= 0; index--) {
                this.remove(comps[index]);
            }
        }

        this.flag.fill(0);
        this.instNames.length = 0;
        this.instNameToComps.clear();
        this.superNameToComps.clear();

        this._node = null;
        this._isValid = false;
        this._world = null;
    }

    /**
     * 销毁
     */
    public destroy() {
        if (this._isDestroyed) return false;
        this._isDestroyed = true;

        if (this.enabled) {
            this.onDisable();
        }

        this.onRemove();

        this._world?.removeEntity(this);

        this.clear();
        return true;
    }

    /**添加*/
    protected onAdd() { }
    /**生效*/
    protected onEnable() { }
    /**失效*/
    protected onDisable() { }
    /**移除*/
    protected onRemove() { }

    /**
     * 检查是否包含全部
     */
    checkFlagAll(flag: number[]) {
        return flagManager.checkFlagAll(this.flag, flag);

    }
    /**
     * 检查是否包含任一
     */
    checkFlagAny(flag: number[]) {
        return flagManager.checkFlagAny(this.flag, flag);
    }

    /**
     * 检查是否仅包含全部
     */
    checkFlagOnly(flag: number[]) {
        const componentNames = this.instNames;
        for (let index = 0; index < componentNames.length; index++) {
            const componentName = componentNames[index];
            const comFlag = flagManager.getAllByName(componentName);
            const success = flagManager.checkFlagAny(comFlag, flag);
            if (!success) {
                return false;
            }
        }
        return true;
    }

    /**
     * 添加相应类型的组件
     * @param param 类型
     */
    add<T extends ITypeofComponent>(param: T | IComponent): InstanceType<T> | null {
        if (DEV && param.ecsName === 'EcsComponent') {
            console.error('请使用ecsclass修饰Component', param);
            return null;
        }

        // 是实例
        if (BaseComponent.isComponent(param)) {
            if (param.entity === this) {
                error(`不允许重复添加同一个组件实例\nuuid:    ${param.uuid}\n组件名:   ${param.ecsName}`);
                return null;
            } else if (param.entity) {
                error(`不允许添加其它实体的组件实例\nuuid:    ${param.uuid}\n组件名:   ${param.ecsName}`);
                return null;
            } else if (!param.allowMultiple) {
                const comps = this.instNameToComps.get(param.ecsName);
                if (comps && comps.length) {
                    error(`此类型的组件不允许重复添加\nuuid:    ${param.uuid}\n组件名:   ${param.ecsName}`);
                    return null;
                }
            }
            this._world.addComponent(this, param);
            return param as InstanceType<T>;
        } else {
            // 如果不允许重复添加
            if (!param.allowMultiple) {
                const comps = this.instNameToComps.get(param.ecsName);
                if (comps && comps.length) {
                    error(`此类型的组件不允许重复添加\n组件名:   ${param.ecsName}`);
                    return null;
                }
            }
            const com = componentCache.get(param.ecsName) || new param();
            this._world.addComponent(this, com);
            return com as InstanceType<T>;
        }
    }

    remove(param: IComponent): boolean {
        if (!param) return false;

        if (param.entity !== this) {
            error(`移除的组件不属于当前实体\n组件名:   ${param.ecsName}`);
            return false;
        }
        this._world.removeComponent(param);
        if (param.allowRecycling) {
            BaseComponent.updateUUID(param as BaseComponent);
            componentCache.put(param.ecsName, param);
        }
        return true;
    }

    removeAll(Com: ITypeofComponent): boolean {
        const comps = this.getAll(Com);
        if (!comps.length) return false;

        return comps.every((com) => {
            return this.remove(com);
        });
    }

    has(Com: ITypeofComponent): boolean {
        const comps = this.getCompsByName(Com.ecsName);
        if (!comps || comps.length === 0) return false;
        return true;
    }

    get<T extends ITypeofComponent>(Com: T): InstanceType<T> | null {
        const comps = this.getCompsByName(Com.ecsName);
        if (!comps || comps.length === 0) return null;
        return comps[0] as InstanceType<T>;
    }

    /**
     * 获取相应类型的组件
     * @param Com 类型
     */
    getAll<T extends ITypeofComponent>(Com: T): InstanceType<T>[] {
        return this.getCompsByName2(Com.ecsName) as InstanceType<T>[];
    }
}
