// @gamex/cc-ecs/assets/ecs/EcsSystem.ts
import { BaseSystem } from '../base/BaseSystem';
import { Filter } from '../core/EcsFilter';
import { ecsclass } from '../core/EcsManager';
import { IEntity, IWorld } from '../core/EcsObject';

@ecsclass('EcsSystem')
export class EcsSystem extends BaseSystem {
    private _world: IWorld | null = null;
    /**获取对应的ecs */
    protected get world() {
        return this._world as IWorld;
    }
    private set world(v) {
        this._world = v;
    }

    protected initialize(world: IWorld) {
        this.world = world;
    }

    /**添加 */
    protected onAdd() { }

    /**移除 */
    protected onRemove() { }

    /**
     * 过滤实体
     * - 配置后才能响应onEntityEnter、onEntityLeave
     */
    protected matcher(): Filter {
        return null;
    }

    /**实体进入系统，会在execute流程最开始调用*/
    protected onEntityEnter(entity: IEntity) { entity; }

    /**实体离开系统，会在execute流程最开始调用*/
    protected onEntityLeave(entity: IEntity) { entity; }

    // 由ecs.execute驱动
    protected execute(dt?: number, ...args: any[]) { dt; args; }
    protected beforeExecute(dt?: number, ...args: any[]) { dt; args; }
    protected afterExecute(dt?: number, ...args: any[]) { dt; args; }

    // 由ecs.update驱动
    protected update(dt?: number, ...args: any[]) { dt; args; }
    protected beforeUpdate(dt?: number, ...args: any[]) { dt; args; }
    protected afterUpdate(dt?: number, ...args: any[]) { dt; args; }

    protected log(...args: any[]) {
        console.log(`[${this.ecsName}] [log]`, ...args);
    }
    protected warn(...args: any[]) {
        console.warn(`[${this.ecsName}] [warn]`, ...args);
    }
    protected error(...args: any[]) {
        console.error(`[${this.ecsName}] [error]`, ...args);
    }
}
