// @gamex/cc-ecs/assets/ecs/EcsWorld.ts
import { Node } from 'cc';
import { DEV } from 'cc/env';
import { BaseEntity } from '../base/BaseEntity';
import { BaseSingleton } from '../base/BaseSingleton';
import { BaseSystem } from '../base/BaseSystem';
import { entityCache } from '../core/EcsCache';
import { Filter } from '../core/EcsFilter';
import { ComponentManager, EntityManager, SystemManager } from '../core/EcsManager';
import { Matcher } from '../core/EcsMatcher';
import { EcsObject, IComponent, IEntity, IEntityUUID, ISingleton, ITypeofComponent, ITypeofEntity, ITypeofSingleton, ITypeofSystem, IWorld } from '../core/EcsObject';

export class EcsWorld implements IWorld {
    private static _inst: EcsWorld;
    public static get inst(): EcsWorld {
        if (!EcsWorld._inst) {
            EcsWorld._inst = new EcsWorld();
        }
        return EcsWorld._inst;
    }

    private entityManager = new EntityManager();
    private systemManager = new SystemManager();
    private componentManager = new ComponentManager();
    private singletons: ISingleton[] = [];
    private matcher: Matcher;

    public constructor() {
        this.matcher = new Matcher(this.entityManager);
    }

    /**
     * 查询实体
     */
    public query<T extends IEntity>(filter: Filter): T[];
    public query<T extends IComponent>(filter: Filter, Component: { new(): T }): T[];
    public query<T>(filter: Filter, Component?: any): T[] {
        const entities = this.matcher.query(filter);
        if (!Component) return entities as T[];
        return entities.map(entity => entity.get(Component as ITypeofComponent)) as T[];
    }

    /**
     * 查询实体
     */
    public find<T extends IEntity>(filter: Filter): T | null;
    public find<T extends IComponent>(filter: Filter, Component: { new(): T }): T | null;
    public find<T>(filter: Filter, Component?: any): T | null {
        const entities = this.matcher.query(filter);
        if (!entities || entities.length === 0) return null;
        if (!Component) return entities[0] as T;
        return entities[0].get(Component as ITypeofComponent) as T;
    }

    /**
     * 查询实体是否存在
     */
    public exist(filter: Filter) {
        const entities = this.matcher.query(filter);
        return !!entities && entities.length > 0;
    }

    /**
     * 根据uuid查询实体
     */
    public findByUuid(uuid: IEntityUUID) {
        return this.entityManager.get(uuid);
    }

    /**
     * 根据uuid查询实体是否存在
     */
    public existByUuid(uuid: IEntityUUID) {
        return !!this.entityManager.has(uuid);
    }

    /**
     * 添加单例组件
     */
    public addSingleton<T extends ITypeofSingleton>(param: T): InstanceType<T>;
    public addSingleton<T extends ISingleton>(param: T): T;
    public addSingleton(param: ITypeofSingleton | ISingleton) {
        if (DEV && param.ecsName === EcsObject.ecsName) {
            console.error('请使用ecsclass修饰', param);
            return null;
        }
        const com = this.singletons.find(com => com.ecsName === param.ecsName);
        if (com) return com;

        if (BaseSingleton.isSingleton(param)) {
            this.singletons.push(param);
            return param;
        } else {
            const com = new param();
            this.singletons.push(com);
            return com;
        }
    }

    /**
     * 获取单例组件
     */
    public getSingleton<T extends ITypeofSingleton>(param: T): InstanceType<T> | null {
        return this.singletons.find(com => com.ecsName === param.ecsName) as InstanceType<T> || null;
    }

    /**
     * 移除单例组件
     */
    public removeSingleton<T extends ITypeofSingleton>(param: T): InstanceType<T> | null {
        const index = this.singletons.findIndex(com => com.ecsName === param.ecsName);
        if (index >= 0) {
            return this.singletons.splice(index, 1)[0] as InstanceType<T>;
        }
        return null;
    }

    /**
     * 创建一个实体
     */
    public createEntity<T extends ITypeofEntity>(Entity: T, node?: Node): InstanceType<T> {
        const entity = entityCache.get() || new Entity();
        BaseEntity.initialize(entity, this, node);
        this.entityManager.add(entity);
        return entity as InstanceType<T>;
    }

    /**
     * 移除一个实体
     */
    public removeEntity(entity: IEntity) {
        this.entityManager.remove(entity);
        if (entity.allowRecycling) {
            BaseEntity.updateUUID(entity);
            entityCache.put(entity);
        }
    }

    public UpdateEntityMatch(entity: IEntity) {
        this.matcher.match(entity);
    }

    /**
     * 添加一个组件
     */
    public addComponent(entity: IEntity, comp: IComponent) {
        this.componentManager.add(comp.ecsName, comp);

        BaseEntity.addComponent(entity, comp);

        this.matcher.match(entity);
    }

    /**
     * 移除一个组件
     */
    public removeComponent(comp: IComponent) {
        this.componentManager.remove(comp.ecsName, comp);

        const entity = comp.entity;
        if (entity) {
            BaseEntity.removeComponent(entity, comp);
            this.matcher.match(entity);
        }
    }

    /**
    * 添加一个系统
    */
    public addSystem<T extends ITypeofSystem>(System: T): void {
        const system = new System();
        BaseSystem.initialize(system, this);
        this.systemManager.add(system);
        BaseSystem.onAdd(system, this.matcher);
    }

    /**
     * 移除一个系统
     */
    // public removeSystem<T extends ITypeofSystem>(System: T): void {
    //     const system = this.systemManager.remove(System);
    //     if (system) BaseSystem.onRemove(system);
    // }

    /**
     * 清理所有数据
     */
    public clear() {
        this.matcher.clear();
        this.systemManager.clear();
        this.entityManager.clear();
        this.componentManager.clear();
        this.singletons.length = 0;
    }

    private executeSystem(dt: number, args: any[]) {
        this.systemManager.each(function (system) {
            BaseSystem.execute(system, dt, args);
        });
    }
    private beforeExecuteSystem(dt: number, args: any[]) {
        this.systemManager.each(function (system) {
            BaseSystem.beforeExecute(system, dt, args);
        });
    }
    private afterExecuteSystem(dt: number, args: any[]) {
        this.systemManager.each(function (system) {
            BaseSystem.afterExecute(system, dt, args);
        });
    }

    private updateSystem(dt: number, args: any[]) {
        this.systemManager.each(function (system) {
            BaseSystem.update(system, dt, args);
        });
    }
    private beforeUpdateSystem(dt: number, args: any[]) {
        this.systemManager.each(function (system) {
            BaseSystem.beforeUpdate(system, dt, args);
        });
    }
    private afterUpdateSystem(dt: number, args: any[]) {
        this.systemManager.each(function (system) {
            BaseSystem.afterUpdate(system, dt, args);
        });
    }

    /**
     * 用于逻辑帧
     * 执行顺序:
     * [系统的beforeExecute] -> [系统的execute] ->[系统的afterExecute]
     */
    public execute(dt: number, ...args: any[]) {
        this.beforeExecuteSystem(dt, args);
        this.executeSystem(dt, args);
        this.afterExecuteSystem(dt, args);
    }

    /**
     * 用于渲染帧
     * 执行顺序:
     * [系统的beforeUpdate] -> [系统的update] -> [系统的afterUpdate]
     */
    public update(dt: number, ...args: any[]) {
        this.beforeUpdateSystem(dt, args);
        this.updateSystem(dt, args);
        this.afterUpdateSystem(dt, args);
    }
}
