// @gamex/cc-ecs/assets/lib/EcsTween.ts
import { math } from 'cc';
import { EcsComponent } from '../ecs/EcsComponent';

// 将Base(类)对应实例中所有值为Type类型的key去除，返回string的复合类型(被过滤后的key)
type KeyofFilterType<Base, ExcludedType> = {
    [key in keyof Base]: Base[key] extends ExcludedType ? never : key;
}[keyof Base];

// 将Base(类)对应实例中所有键为Key类型的key去除，返回string的复合类型(被过滤后的key)
type KeyofFilterKey<Base, ExcludedKey> = {
    [key in keyof Base]: key extends ExcludedKey ? never : key;
}[keyof Base];

type ExcludedKey = keyof EcsComponent; //| keyof INodeReadonly;
type PropType<Base> = Partial<
    Pick<Base, KeyofFilterType<
        Pick<Base, KeyofFilterKey<Base, ExcludedKey>>,
        Function | string | object | boolean | symbol
    >>
>;

type IObject = { [key: string]: any };

/** 缓动函数，可以使用已有的，也可以传入自定义的函数。 */
type IEasing = (k: number) => number;

enum TweenActionType {
    None,
    To,
    By
}

// 动作基类
abstract class TweenAction {
}

// 属性动作
class PropTweenAction extends TweenAction {
    type: TweenActionType = TweenActionType.None;
    /**进度 */
    ratio: number = 0;
    /**时间 */
    time: number = 0;
    /**结束属性 */
    end: { [key: string]: any } | null = null;
    /**开始属性 */
    start: { [key: string]: any } | null = null;
    /**缓动函数 */
    easing: IEasing | null = null;
}

// 贝塞尔属性动作
class BezierPropTweenAction extends PropTweenAction {
    /**控制位属性 */
    control: { [key: string]: any } | null = null;
}

// 等待动作
class DelayTweenAction extends TweenAction {
    delay: number = 0;
}

// 回调动作
class CallTweenAction extends TweenAction {
    callback: Function | null = null;
}

/**
 * 缓动任务
 */
class Task<T extends EcsComponent> {
    static step(tween: Task<EcsComponent>, dt: number) {
        // 如果没有动作，则移除
        if (tween.actions.length == 0) return false;
        // 如果目标对象无效，则移除
        if (tween.target.isValid == false) return false;
        // 如果目标对象被销毁(Uuid不一致说明是被销毁后复用了)，则移除
        if (tween.target.uuid != tween.targetUuid) return false;

        const action = tween.actions[0];
        if (action instanceof DelayTweenAction) {
            action.delay -= dt;
            if (action.delay <= 0) {
                tween.actions.shift();
            }
        } else if (action instanceof CallTweenAction) {
            if (action.callback) {
                tween.actions.shift();
                action.callback();
            }
        } else if (action instanceof PropTweenAction) {
            // 整体进度
            action.ratio = Math.min(action.ratio + dt / action.time, 1);
            // 完成移除
            if (action.ratio >= 1) {
                tween.actions.shift();
            }
            // 应用缓动函数后的进度
            const ratio = action.easing ? action.easing(action.ratio) : action.ratio;
            // 计算动态数值
            if (!action.start) {
                // 初始化起点坐标
                action.start = Task.getStartProps(action.end, tween.target);
                if (action.type === TweenActionType.By) {
                    if (action instanceof BezierPropTweenAction) {
                        action.end = Task.getEndProps(action.end, tween.target);
                        action.control = Task.getEndProps(action.control, tween.target);
                    } else {
                        action.end = Task.getEndProps(action.end, tween.target);
                    }
                }
            }
            // 插值运算
            if (action instanceof BezierPropTweenAction) {
                Task.bezier(ratio, action.start, action.control, action.end, tween.target);
            } else {
                Task.lerp(ratio, action.start, action.end, tween.target);
            }
        } else {
            tween.actions.shift();
        }

        return tween.actions.length > 0;
    }

    private static Tid = 0;

    /**唯一id */
    private _id = ++Task.Tid;
    public get id() {
        return this._id;
    }

    /**动作列表 */
    private actions: TweenAction[] = [];

    /**目标对象 */
    private targetUuid: number;
    private _target!: T;
    public get target(): T {
        return this._target;
    }

    constructor(target: T) {
        this._target = target;
        this.targetUuid = target.uuid;
    }

    /**
     * 等待
     * @param delay
     * @returns 
     */
    delay(delay: number) {
        const action = new DelayTweenAction();
        action.delay = delay;
        this.actions.push(action);
        return this;
    }

    /**
     * 执行
     * @param callback 
     * @returns 
     */
    call(callback: Function) {
        const action = new CallTweenAction();
        action.callback = callback;
        this.actions.push(action);
        return this;
    }

    /**
     * 绝对缓动
     * @param time 时间
     * @param props 属性(只能是简单类型)
     * @param opts 选项
     * @returns 
     */
    to(time: number, props: PropType<T>, easing: IEasing = null) {
        const action = new PropTweenAction();
        action.type = TweenActionType.To;
        action.time = Math.max(time, 0);
        action.end = props;
        action.start = null;// 运行时动态计算
        action.easing = easing;
        this.actions.push(action);
        return this;
    }

    /**
     * 相对缓动
     * @param time 时间
     * @param props 属性(只能是简单类型)
     * @param opts 选项
     * @returns 
     */
    by<T extends EcsComponent>(time: number, props: PropType<T>, easing: IEasing = null) {
        const action = new PropTweenAction();
        action.type = TweenActionType.By;
        action.time = Math.max(time, 0);
        action.start = null;// 运行时动态计算
        action.end = props;// 运行时动态计算(此处用作临时存储)
        action.easing = easing;
        this.actions.push(action);
        return this;
    }

    /**
     * 绝对贝塞尔缓动
     * @param time 时间
     * @param target 目标
     * @param props 属性(只能是简单类型)
     * @param control 控制值(影响曲线幅度)
     * @param opts 选项
     * @returns 
     */
    bezierTo<T extends EcsComponent, K extends PropType<T>>(time: number, props: K, control: Required<K>, easing: IEasing = null) {
        const action = new BezierPropTweenAction();
        action.type = TweenActionType.To;
        action.time = Math.max(time, 0);
        action.start = null;// 运行时动态计算
        action.control = control;
        action.end = props;
        action.easing = easing;
        this.actions.push(action);
        return this;
    }

    /**
     * 相对贝塞尔缓动
     * @param time 时间
     * @param target 目标
     * @param props 属性(只能是简单类型)
     * @param control 控制值(影响曲线幅度)
     * @param opts 选项
     */
    bezierBy<T extends EcsComponent, K extends PropType<T>>(time: number, props: K, control: Required<K>, easing: IEasing = null) {
        const action = new BezierPropTweenAction();
        action.type = TweenActionType.By;
        action.time = Math.max(time, 0);
        action.start = null;// 运行时动态计算
        action.control = control;// 运行时动态计算
        action.end = props;// 运行时动态计算
        action.easing = easing;
        this.actions.push(action);
        return this;
    }

    /**
     * 获取目标值
     * @param props 差值
     * @param target 原始值
     */
    private static getEndProps(props: IObject, target: IObject) {
        const end: IObject = {};

        for (const key in props) {
            if (Object.prototype.hasOwnProperty.call(props, key)) {
                if (typeof props[key] === 'number') {
                    end[key] = props[key] + (target[key] || 0);
                } else if (typeof props[key] === 'boolean') {
                    end[key] = props[key];
                } else if (props[key] && typeof props[key] === 'object') {
                    end[key] = this.getEndProps(props[key], target[key]);
                }
            }
        }
        return end;
    }

    /**
     * 获取初始值
     * @param props 作为过滤使用
     * @param target 原始值
     */
    private static getStartProps(props: IObject, target: IObject) {
        const start: IObject = {};

        for (const key in props) {
            if (Object.prototype.hasOwnProperty.call(props, key)) {
                if (typeof props[key] === 'number') {
                    start[key] = target[key] || 0;
                } else if (typeof props[key] === 'boolean') {
                    start[key] = target[key] || false;
                } else if (props[key] && typeof props[key] === 'object') {
                    start[key] = this.getStartProps(props[key], target[key]);
                }
            }
        }
        return start;
    }

    private static lerp(ratio: number, start: IObject, end: IObject, target: IObject) {
        for (const key in start) {
            if (Object.prototype.hasOwnProperty.call(start, key)) {
                if (typeof start[key] === 'number') {
                    target[key] = math.lerp(start[key], end[key], ratio);
                } else if (typeof start[key] === 'boolean') {
                    target[key] = ratio >= 1 ? end[key] : start[key];
                } else if (start[key] && typeof start[key] === 'object') {
                    this.lerp(ratio, start[key], end[key], target[key]);
                }
            }
        }
    }

    private static bezier(ratio: number, start: IObject, control: IObject, end: IObject, target: IObject) {
        for (const key in start) {
            if (Object.prototype.hasOwnProperty.call(start, key)) {
                if (typeof start[key] === 'number') {
                    const n1 = math.lerp(start[key], control[key], ratio);
                    const n2 = math.lerp(control[key], end[key], ratio);
                    target[key] = math.lerp(n1, n2, ratio);
                } else if (typeof start[key] === 'boolean') {
                    target[key] = ratio >= 1 ? end[key] : start[key];
                } else if (start[key] && typeof start[key] === 'object') {
                    this.bezier(ratio, start[key], control[key], end[key], target[key]);
                }
            }
        }
    }
}

/**
 * 缓动
 * @example
 * // 创建tween
 * const tween = new Tween();
 * tween.add(组件实例)
 *      .delay(1);
 *      .by(1, ...);
 * // 迭代一次
 * tween.step(dt);
 */
export class Tween {
    public static Task = Task;

    private tasks = new Map<number, Task<EcsComponent>>();
    private pausing = new Map<number, true>();

    /**
     * 添加tween
     * @param target 
     * @returns 
     */
    add<T extends EcsComponent>(target: T) {
        let task = new Task(target);
        this.tasks.set(task.id, task);
        return task;
    }

    /**
     * 移除指定的tween
     * @param id 
     */
    remove(id: number): void;
    /**
     * 移除指定的tween
     * @param tween 
     */
    remove(tween: Task<EcsComponent>): void;
    /**
     * 移除指定target的所有tween
     * @param target 
     */
    remove(target: EcsComponent): void;
    remove(param: number | EcsComponent | Task<EcsComponent>) {
        if (typeof param === 'number') {
            this.tasks.delete(param);
            this.pausing.delete(param);
        } else if (param instanceof Task) {
            this.tasks.delete(param.id);
            this.pausing.delete(param.id);
        } else {
            this.tasks.forEach((tween, tid) => {
                if (tween.target === param) {
                    this.tasks.delete(tid);
                    this.pausing.delete(tid);
                }
            });
        }
    }

    /**
     * 暂停指定的tween
     * @param id 
     */
    pause(id: number): void;
    /**
     * 暂停指定的tween
     * @param tween 
     */
    pause(tween: Task<EcsComponent>): void;
    /**
     * 暂停指定target的所有tween
     * @param target 
     */
    pause(target: EcsComponent): void;
    pause(param: number | EcsComponent | Task<EcsComponent>) {
        if (typeof param === 'number') {
            this.pausing.set(param, true);
        } else if (param instanceof Task) {
            this.pausing.set(param.id, true);
        } else {
            this.tasks.forEach((tween, tid) => {
                if (tween.target === param) {
                    this.pausing.set(tid, true);
                }
            });
        }
    }

    /**
     * 恢复指定的tween
     * @param id 
     */
    resume(id: number): void;
    /**
     * 恢复指定的tween
     * @param tween 
     */
    resume(tween: Task<EcsComponent>): void;
    /**
     * 恢复指定target的所有tween
     * @param target 
     */
    resume(target: EcsComponent): void;
    resume(param: number | EcsComponent | Task<EcsComponent>) {
        if (typeof param === 'number') {
            this.pausing.delete(param);
        } else if (param instanceof Task) {
            this.pausing.delete(param.id);
        } else {
            this.tasks.forEach((tween, tid) => {
                if (tween.target === param) {
                    this.pausing.delete(tid);
                }
            });
        }
    }

    /**
     * 迭代一次
     */
    step(dt: number) {
        this.tasks.forEach((tween, tid) => {
            // 暂停中，则跳过
            if (this.pausing.has(tid)) return;
            // 执行次数达到上限，则移除
            if (Task.step(tween, dt) === false) {
                this.tasks.delete(tid);
            }
        });
    }

    /**
     * 清空
     */
    clear() {
        this.tasks.clear();
        this.pausing.clear();
    }

    get size() {
        return this.tasks.size;
    }
}
