// @gamex/cc-ecs/assets/lib/EcsTimer.ts
import { EcsObject } from '../core/EcsObject';

interface ICallback {
    (...args: any[]): any
}

/**
 * 定时器任务
 */
class Task {
    static step(timer: Task, dt: number) {
        timer._time += dt;

        if (timer._first) {
            if (timer._time < timer._delay) return true;
            timer._first = false;
            timer._time = 0;
        } else if (timer._time < timer._interval) {
            return true;
        } else {
            timer._time = 0;
        }

        timer.callback.call(timer.target);

        // 执行次数达到上限，则移除
        if (timer._repeat >= 0 && --timer._repeat < 0) {
            return false;
        }

        return true;
    }

    private static Tid = 0;

    /**定时器ID */
    private _id = ++Task.Tid;
    public get id() {
        return this._id;
    }

    /**是否执行过 */
    private _first = true;
    /**执行计时 */
    private _time = 0;
    /**首次执行的延迟时间 */
    private _delay = 0;
    /**重复执行的间隔时间 */
    private _interval = 0;
    /**重复执行次数(<0为无限次) */
    private _repeat = 0;

    /**回调函数 */
    private callback!: ICallback;
    private _target: EcsObject | null = null;
    public get target(): EcsObject | null {
        return this._target;
    }

    constructor(callback: ICallback, target: EcsObject | null = null) {
        this._target = target;
        this.callback = callback;
    }

    /**
     * 首次执行延迟时间
     * @param value 
     * @returns 
     */
    delay(value: number) {
        this._delay = Math.max(0, value);
        return this;
    }

    /**
     * 重复执行间隔时间
     * @param value 
     * @returns 
     */
    interval(value: number) {
        this._interval = Math.max(0, value);
        return this;
    }

    /**
     * 重复执行次数(<0为无限次)
     * - 初始值为0
     * - 实际执行次数为repeat + 1
     * @param value 
     * @returns 
     */
    repeat(value: number) {
        this._repeat = value;
        return this;
    }
}

/**
 * 定时器管理器
 * @example
 * // 创建timer
 * const timer = new Timer();
 * timer.add(() => { consoe.log('timer'); }).repeat(10);
 * // 迭代一次
 * timer.step(dt);
 */
export class Timer {
    public static Task = Task;

    private timers = new Map<number, Task>();
    private pausing = new Map<number, true>();

    /**
     * 添加定时器
     * @param callback 回调函数
     * @param target 绑定的目标
     * @returns 
     */
    add(callback: ICallback, target: EcsObject | null = null) {
        let time = new Task(callback, target);
        this.timers.set(time.id, time);
        return time;
    }

    /**
     * 移除指定的定时器
     * @param id 
     */
    remove(id: number): void;
    /**
     * 移除指定的定时器
     * @param timer 
     */
    remove(timer: Task): void;
    /**
     * 移除指定的定时器
     * @param target 
     */
    remove(target: EcsObject): void;
    remove(param: number | Task | EcsObject) {
        if (typeof param === 'number') {
            this.timers.delete(param);
            this.pausing.delete(param);
        } else if (param instanceof Task) {
            this.timers.delete(param.id);
            this.pausing.delete(param.id);
        } else {
            this.timers.forEach((tween, tid) => {
                if (tween.target === param) {
                    this.timers.delete(tid);
                    this.pausing.delete(tid);
                }
            });
        }
    }

    /**
     * 暂停指定的定时器
     * @param id 
     */
    pause(id: number): void;
    /**
     * 暂停指定的定时器
     * @param timer 
     */
    pause(timer: Task): void;
    /**
     * 暂停指定的定时器
     * @param target 
     */
    pause(target: EcsObject): void;
    pause(param: number | Task | EcsObject) {
        if (typeof param === 'number') {
            this.pausing.set(param, true);
        } else if (param instanceof Task) {
            this.pausing.set(param.id, true);
        } else {
            this.timers.forEach((tween, tid) => {
                if (tween.target === param) {
                    this.pausing.set(tid, true);
                }
            });
        }
    }

    /**
     * 恢复指定的定时器
     * @param id 
     */
    resume(id: number): void;
    /**
     * 恢复指定的定时器
     * @param timer 
     */
    resume(timer: Task): void;
    /**
     * 恢复指定的定时器
     * @param target 
     */
    resume(target: EcsObject): void;
    resume(param: number | Task | EcsObject) {
        if (typeof param === 'number') {
            this.pausing.delete(param);
        } else if (param instanceof Task) {
            this.pausing.delete(param.id);
        } else {
            this.timers.forEach((tween, tid) => {
                if (tween.target === param) {
                    this.pausing.delete(tid);
                }
            });
        }
    }

    /**
     * 迭代一次
     */
    step(dt: number) {
        this.timers.forEach((timer, tid) => {
            // 暂停中，则跳过
            if (this.pausing.has(tid)) return;
            // 执行次数达到上限，则移除
            if (Task.step(timer, dt) === false) {
                this.timers.delete(tid);
            }
        });
    }

    /**
     * 清空
     */
    clear() {
        this.timers.clear();
        this.pausing.clear();
    }

    get size() {
        return this.timers.size;
    }
}
