// @gamex/cc-ecs/assets/core/EcsFilter.ts
import { flagManager } from './EcsManager';
import { IEntity, IFlag, ITypeofComponent } from './EcsObject';

/**
 * 过滤器
 */
export class Filter {
    static check(filter: Filter, entity: IEntity) {
        if (entity.enabled == false) return false;
        if (entity.isValid == false) return false;

        for (let index = 0; index < filter.pipeline.length; index++) {
            if (filter.pipeline[index](entity) == false) {
                return false;
            }
        }
        return true;
    }

    static clone(filter: Filter) {
        const clone = new Filter();
        Array.prototype.push.apply(clone.allList, filter.allList);
        Array.prototype.push.apply(clone.anyList, filter.anyList);
        Array.prototype.push.apply(clone.onlyList, filter.onlyList);
        Array.prototype.push.apply(clone.excludeList, filter.excludeList);
        Array.prototype.push.apply(clone.pipeline, filter.pipeline);
        return clone;
    }

    private pipeline: ((entity: IEntity) => boolean)[] = [];

    private allList: ITypeofComponent[] = [];
    private anyList: ITypeofComponent[] = [];
    private onlyList: ITypeofComponent[] = [];
    private excludeList: ITypeofComponent[] = [];

    private _allFlag: IFlag = null;
    private get allFlag() {
        if (!this._allFlag) {
            this._allFlag = flagManager.getByNames(this.allList.map(Com => Com.ecsName));
        }
        return this._allFlag;
    }
    private _anyFlag: IFlag = null;
    private get anyFlag() {
        if (!this._anyFlag) {
            this._anyFlag = flagManager.getByNames(this.anyList.map(Com => Com.ecsName));
        }
        return this._anyFlag;
    }
    private _onlyFlag: IFlag = null;
    private get onlyFlag() {
        if (!this._onlyFlag) {
            this._onlyFlag = flagManager.getByNames(this.onlyList.map(Com => Com.ecsName));
        }
        return this._onlyFlag;
    }
    private _excludeFlag: IFlag = null;
    private get excludeFlag() {
        if (!this._excludeFlag) {
            this._excludeFlag = flagManager.getByNames(this.excludeList.map(Com => Com.ecsName));
        }
        return this._excludeFlag;
    }

    private _key = '';
    public get key() {
        if (!this._key) {
            this._key = this.allFlag.join(',') + '-' + this.anyFlag.join(',') + '-' + this.onlyFlag.join(',') + '-' + this.excludeFlag.join(',');
        }
        return this._key;
    }

    /**
     * 必须包含所有这些组件
     */
    all(...Comps: ITypeofComponent[]) {
        Array.prototype.push.apply(this.allList, Comps);

        this.pipeline.push((entity: IEntity) => {
            return entity.checkFlagAll(this.allFlag) && this.allList.every(Com => {
                return entity.getAll(Com).some(comp => comp.enabled && comp.isValid);
            });
        });
        return this;
    }
    /**
     * 有这些组件中的任何一个
     */
    any(...Comps: ITypeofComponent[]) {
        Array.prototype.push.apply(this.anyList, Comps);

        this.pipeline.push((entity: IEntity) => {
            return entity.checkFlagAny(this.anyFlag) && this.anyList.some(Com => {
                return entity.getAll(Com).some(comp => comp.enabled && comp.isValid);
            });
        });
        return this;
    }
    /**
     * 仅仅只有这些组件
     */
    only(...Comps: ITypeofComponent[]) {
        Array.prototype.push.apply(this.onlyList, Comps);

        this.pipeline.push((entity: IEntity) => {
            return entity.checkFlagOnly(this.onlyFlag) && this.onlyList.every(Com => {
                return entity.getAll(Com).some(comp => comp.enabled && comp.isValid);
            });
        });
        return this;
    }
    /**
     * 不能包含其中的任何一个组件
     */
    exclude(...Comps: ITypeofComponent[]) {
        Array.prototype.push.apply(this.excludeList, Comps);

        this.pipeline.push((entity: IEntity) => {
            return !entity.checkFlagAny(this.excludeFlag) || this.excludeList.every(Com => {
                return entity.getAll(Com).every(comp => !comp.enabled || !comp.isValid);
            });
        });
        return this;
    }
}

/**
 * 过滤条件
 */
export const filter = {
    /**
     * 必须包含所有这些组件
     */
    all(...comps: ITypeofComponent[]) {
        return new Filter().all(...comps);
    },
    /**
     * 有这些组件中的任何一个
     */
    any(...comps: ITypeofComponent[]) {
        return new Filter().any(...comps);
    },
    /**
     * 仅仅只有这些组件
     */
    only(...comps: ITypeofComponent[]) {
        return new Filter().only(...comps);
    },
    /**
     * 不能包含其中的任何一个组件
     */
    exclude(...comps: ITypeofComponent[]) {
        return new Filter().exclude(...comps);
    }
};
