// @gamex/cc-ecs/assets/core/EcsMatcher.ts
// 匹配器，用于根据Filter给Entity进行分组
// 实体每次添加组件时都要经过这里，World中的query操作也是从这里直接取值
// 每个World都有自己的Matcher实例

import { Filter } from './EcsFilter';
import { EntityManager } from './EcsManager';
import { IEntity } from './EcsObject';

type IListener = (entity: IEntity) => void;

class Group {
    private entities = new Map<number, IEntity>();
    private onEntityEnterListeners: IListener[] = [];
    private onEntityLeaveListeners: IListener[] = [];
    constructor(private filter: Filter) {
    }

    public onEntityEnter(listener: IListener) {
        this.onEntityEnterListeners.push(listener);
    }
    public onEntityLeave(listener: IListener) {
        this.onEntityLeaveListeners.push(listener);
    }

    public match(entity: IEntity) {
        if (Filter.check(this.filter, entity)) {
            if (!this.entities.has(entity.uuid)) {
                this.entities.set(entity.uuid, entity);
                for (let index = 0; index < this.onEntityEnterListeners.length; index++) {
                    const listener = this.onEntityEnterListeners[index];
                    listener(entity);
                }
            }
        } else {
            if (this.entities.has(entity.uuid)) {
                this.entities.delete(entity.uuid);
                for (let index = 0; index < this.onEntityLeaveListeners.length; index++) {
                    const listener = this.onEntityLeaveListeners[index];
                    listener(entity);
                }
            }
        }
    }

    public getEntities() {
        return Array.from(this.entities.values());
    }
}

export class Matcher {
    static Group = Group;

    private groups = new Map<string, Group>();
    constructor(private entityManager: EntityManager) {
    }

    public clear() {
        this.groups.clear();
    }

    public match(entity: IEntity) {
        this.groups.forEach(group => group.match(entity));
    }

    public register(filter: Filter) {
        if (this.groups.has(filter.key)) {
            return this.groups.get(filter.key);
        }
        else {
            const group = new Group(Filter.clone(filter));
            this.groups.set(filter.key, group);
            this.entityManager.each(entity => group.match(entity));
            return group;
        }
    }

    public query(filter: Filter) {
        const group = this.register(filter);
        return group.getEntities();
    }
}
