// @gamex/cc-ecs/assets/core/EcsObject.ts
import { Node } from 'cc';
import { Filter } from './EcsFilter';

export abstract class EcsObject {
    /**类名 */
    static ecsName = 'EcsObject';
    /**类名 */
    public get ecsName() {
        return (this.constructor as typeof EcsObject).ecsName;
    }
}

export type IFlag = number[];

export type IEntityUUID = number;
export type IComponentName = string;
export type IComponentUUID = number;

export type ITypeofEntity = { new(): IEntity, ecsName: string, allowRecycling: boolean };
export type ITypeofComponent = { new(): IComponent, ecsName: string, allowRecycling: boolean, allowMultiple: boolean };
export type ITypeofSystem = { new(): ISystem, ecsName: string };
export type ITypeofSingleton = { new(): ISingleton, ecsName: string };

export interface IEntity {
    /**
     * 唯一Id
     */
    readonly uuid: number
    /**
     * 唯一名
     */
    readonly ecsName: string
    /**
     * 允许回收复用(默认false)
     */
    readonly allowRecycling: boolean
    /**
     * 是否有效
    */
    readonly isValid: boolean

    /**
     * 绑定的节点
     */
    node: Node | null
    /**
     * 标记是否生效(为false后将不会被查询到)
     */
    enabled: boolean
    /**
     * 销毁
     * 会在触发onDisable和移除所有组件之后才会isValid设为false
     */
    destroy(): boolean
    /**
     * 检查是否仅包含全部
     */
    checkFlagAll(flag: IFlag): boolean
    /**
     * 检查是否包含任一
     */
    checkFlagAny(flag: IFlag): boolean
    /**
     * 检查是否仅包含全部
     */
    checkFlagOnly(flag: IFlag): boolean
    /**
     * 添加相应类型的组件
     * @param param 类型
     * @returns 
     */
    add<T extends ITypeofComponent>(Com: T | IComponent): InstanceType<T> | null
    /**
     * 有没有相应类型的组件
     */
    has(Com: ITypeofComponent): boolean
    /**
     * 获取相应类型的组件
     */
    get<T extends ITypeofComponent>(Com: T): InstanceType<T> | null
    /**
     * 获取相应类型的组件
     */
    getAll<T extends ITypeofComponent>(Com: T): InstanceType<T>[]
    /**
     * 移除组件
     * @param com 实例
     * @returns 
     */
    remove(com: IComponent): boolean;
    /**
     * 移除相应组件的所有组件
     * @param Com 类型
     * @returns 
     */
    removeAll(Com: ITypeofComponent): boolean;
}

export interface IComponent<E extends IEntity = IEntity> {
    /**
     * 唯一Id
     */
    readonly uuid: number
    /**
     * 唯一名
     */
    readonly ecsName: string
    /**
     * 允许回收复用(默认false)
     */
    readonly allowRecycling: boolean
    /**
     * 允许在同一个实体中重复添加(默认false)
     */
    readonly allowMultiple: boolean

    /**是否有效*/
    readonly isValid: boolean

    /**
     * 标记是否生效(为false代表不会被查询到)
     */
    enabled: boolean
    /**
     * 表示该组件是否被启用并且所在的实体也处于启用状态(为false代表不会被查询到)
     */
    readonly enabledInHierarchy: boolean
    /**当前组件所在的实体 */
    readonly entity: E | null;
    /**销毁当前组件 */
    destroy(): boolean;
}

export interface ISystem {
    /**
     * 唯一Id
     */
    readonly uuid: number
    /**
     * 唯一名
     */
    readonly ecsName: string
}

export type ISingleton = {
    /**
     * 唯一Id
     */
    readonly uuid: number
    /**
     * 唯一名
     */
    readonly ecsName: string
};

export interface IWorld {
    /**
     * 查询实体
     */
    query<T extends IEntity>(filter: Filter): T[];
    query<T extends IComponent>(filter: Filter, Component: { new(): T }): T[];
    query<T>(filter: Filter, Component?: ITypeofComponent): T[];
    /**
     * 查询实体
     */
    find<T extends IEntity>(filter: Filter): T;
    find<T extends IComponent>(filter: Filter, Component: { new(): T }): T;
    find<T>(filter: Filter, Component?: ITypeofComponent): T;
    /**
     * 查询实体是否存在
     */
    exist(filter: Filter): boolean;
    /**
     * 根据uuid查询实体
     */
    findByUuid(uuid: IEntityUUID): IEntity | null;
    /**
     * 根据uuid查询实体是否存在
     */
    existByUuid(uuid: IEntityUUID): boolean;
    /**
     * 添加系统
     */
    addSystem<T extends ITypeofSystem>(System: T): void
    /**
     * 移除系统
     */
    // removeSystem<T extends ITypeofSystem>(System: T): void
    /**
     * 添加单例组件
     */
    addSingleton<T extends ITypeofSingleton>(param: T): InstanceType<T>
    /**
     * 添加单例组件
     */
    addSingleton<T extends ISingleton>(param: T): T
    /**
     * 获取单例组件
     */
    getSingleton<T extends ITypeofSingleton>(param: T): InstanceType<T> | null
    /**
     * 移除单例组件
     */
    removeSingleton<T extends ITypeofSingleton>(param: T): InstanceType<T> | null
    /**
     * 创建一个实体
     */
    createEntity<T extends ITypeofEntity>(Entity: T, options?: Node | { node?: Node }): InstanceType<T>
    /**
     * 清理系统、实体、组件、单例组件(都不会触发生命周期)
     */
    clear(): any
    /**
     * 用于逻辑帧
     * 执行顺序:
     * [系统的beforeExecute] -> [系统的execute] ->[系统的afterExecute]
     */
    execute(dt: number, ...args: any[]): any
    /**
     * 用于渲染帧
     * 执行顺序:
     * [系统的beforeUpdate] -> [系统的update] -> [系统的afterUpdate]
     */
    update(dt: number, ...args: any[]): any
}
